File: medbrains/crates/medbrains-core/src/command_center.rs
//! Command Center domain types — patient flow, department load, alerts, transport, KPIs.

use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

// ── Patient Flow ────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct PatientFlowSnapshot {
    pub registered_today: i64,
    pub opd_waiting: i64,
    pub opd_in_consult: i64,
    pub er_active: i64,
    pub ipd_admitted: i64,
    pub pending_discharge: i64,
    pub discharged_today: i64,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct HourlyFlowRow {
    pub hour: i32,
    pub admissions: i64,
    pub discharges: i64,
    pub er_arrivals: i64,
    pub opd_visits: i64,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct BottleneckRow {
    pub area: String,
    pub metric: String,
    pub current_value: f64,
    pub threshold: f64,
    pub severity: String,
}

// ── Department Load ─────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct DepartmentLoadRow {
    pub department_id: Uuid,
    pub department_name: String,
    pub bed_total: i64,
    pub bed_occupied: i64,
    pub occupancy_pct: f64,
    pub queue_depth: i64,
    pub avg_wait_mins: f64,
}

// ── Alerts ──────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct DepartmentAlertRow {
    pub id: Uuid,
    pub department_id: Uuid,
    pub department_name: String,
    pub alert_level: String,
    pub metric_code: String,
    pub current_value: Decimal,
    pub threshold_value: Decimal,
    pub message: String,
    pub acknowledged_by: Option<Uuid>,
    pub acknowledged_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct AlertThresholdRow {
    pub id: Uuid,
    pub department_id: Uuid,
    pub department_name: String,
    pub metric_code: String,
    pub warning_threshold: Option<Decimal>,
    pub critical_threshold: Option<Decimal>,
    pub is_active: bool,
}

#[derive(Debug, Deserialize)]
pub struct CreateAlertThresholdRequest {
    pub department_id: Uuid,
    pub metric_code: String,
    pub warning_threshold: Option<f64>,
    pub critical_threshold: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateAlertThresholdRequest {
    pub warning_threshold: Option<f64>,
    pub critical_threshold: Option<f64>,
    pub is_active: Option<bool>,
}

// ── Discharge Coordinator ───────────────────────────────

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct PendingDischargeRow {
    pub admission_id: Uuid,
    pub patient_id: Uuid,
    pub patient_name: String,
    pub uhid: String,
    pub ward_name: String,
    pub bed_code: String,
    pub doctor_name: String,
    pub admitted_at: DateTime<Utc>,
    pub expected_discharge_date: Option<NaiveDate>,
    pub days_admitted: i32,
    pub pending_labs: i64,
    pub pending_billing: bool,
    pub summary_draft: bool,
    pub checklist_pending: i64,
}

// ── Bed Turnaround / Environmental Services ─────────────

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct BedTurnaroundRow {
    pub location_id: Uuid,
    pub location_code: String,
    pub ward_name: String,
    pub status: String,
    pub discharge_at: Option<DateTime<Utc>>,
    pub cleaning_started_at: Option<DateTime<Utc>>,
    pub cleaning_completed_at: Option<DateTime<Utc>>,
    pub turnaround_minutes: Option<i32>,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct TurnaroundStatsRow {
    pub ward_name: String,
    pub avg_turnaround_mins: f64,
    pub max_turnaround_mins: i32,
    pub beds_awaiting_cleaning: i64,
    pub beds_being_cleaned: i64,
}

// ── Transport ───────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct TransportRequestRow {
    pub id: Uuid,
    pub patient_name: Option<String>,
    pub from_location: Option<String>,
    pub to_location: Option<String>,
    pub transport_mode: String,
    pub status: String,
    pub priority: String,
    pub requested_by_name: String,
    pub assigned_to_name: Option<String>,
    pub requested_at: DateTime<Utc>,
    pub assigned_at: Option<DateTime<Utc>>,
    pub picked_up_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateTransportRequest {
    pub patient_id: Option<Uuid>,
    pub admission_id: Option<Uuid>,
    pub from_location_id: Option<Uuid>,
    pub to_location_id: Option<Uuid>,
    pub transport_mode: String,
    pub priority: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateTransportRequest {
    pub transport_mode: Option<String>,
    pub priority: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AssignTransportRequest {
    pub assigned_to: Uuid,
}

// ── KPIs ────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize)]
pub struct KpiTile {
    pub code: String,
    pub label: String,
    pub value: f64,
    pub unit: String,
    pub trend: Option<f64>,
    pub period: String,
}

File: medbrains/packages/types/src/command-center.ts
// Command center types — split from index.ts, barrel-re-exported.

// ── Command Center ─────────────────────────────────────────

export interface PatientFlowSnapshot {
  registered_today: number;
  opd_waiting: number;
  opd_in_consult: number;
  er_active: number;
  ipd_admitted: number;
  pending_discharge: number;
  discharged_today: number;
}

export interface HourlyFlowRow {
  hour: number;
  admissions: number;
  discharges: number;
  er_arrivals: number;
  opd_visits: number;
}

export interface BottleneckRow {
  area: string;
  metric: string;
  current_value: number;
  threshold: number;
  severity: string;
}

export interface DepartmentLoadRow {
  department_id: string;
  department_name: string;
  bed_total: number;
  bed_occupied: number;
  occupancy_pct: number;
  queue_depth: number;
  avg_wait_mins: number;
}

export interface DepartmentAlertRow {
  id: string;
  department_id: string;
  department_name: string;
  alert_level: string;
  metric_code: string;
  current_value: number;
  threshold_value: number;
  message: string;
  acknowledged_by: string | null;
  acknowledged_at: string | null;
  created_at: string;
}

export interface AlertThresholdRow {
  id: string;
  department_id: string;
  department_name: string;
  metric_code: string;
  warning_threshold: number | null;
  critical_threshold: number | null;
  is_active: boolean;
}

export interface CreateAlertThresholdRequest {
  department_id: string;
  metric_code: string;
  warning_threshold?: number;
  critical_threshold?: number;
}

export interface UpdateAlertThresholdRequest {
  warning_threshold?: number;
  critical_threshold?: number;
  is_active?: boolean;
}

export interface PendingDischargeRow {
  admission_id: string;
  patient_id: string;
  patient_name: string;
  uhid: string;
  ward_name: string;
  bed_code: string;
  doctor_name: string;
  admitted_at: string;
  expected_discharge_date: string | null;
  days_admitted: number;
  pending_labs: number;
  pending_billing: boolean;
  summary_draft: boolean;
  checklist_pending: number;
}

export interface BedTurnaroundRow {
  location_id: string;
  location_code: string;
  ward_name: string;
  status: string;
  discharge_at: string | null;
  cleaning_started_at: string | null;
  cleaning_completed_at: string | null;
  turnaround_minutes: number | null;
}

export interface TurnaroundStatsRow {
  ward_name: string;
  avg_turnaround_mins: number;
  max_turnaround_mins: number;
  beds_awaiting_cleaning: number;
  beds_being_cleaned: number;
}

export interface TransportRequestRow {
  id: string;
  patient_name: string | null;
  from_location: string | null;
  to_location: string | null;
  transport_mode: string;
  status: string;
  priority: string;
  requested_by_name: string;
  assigned_to_name: string | null;
  requested_at: string;
  assigned_at: string | null;
  picked_up_at: string | null;
  completed_at: string | null;
  notes: string | null;
}

export interface CreateTransportRequest {
  patient_id?: string;
  admission_id?: string;
  from_location_id?: string;
  to_location_id?: string;
  transport_mode: string;
  priority?: string;
  notes?: string;
}

export interface UpdateTransportRequest {
  transport_mode?: string;
  priority?: string;
  notes?: string;
}

export interface AssignTransportRequest {
  assigned_to: string;
}

export interface KpiTile {
  code: string;
  label: string;
  value: number;
  unit: string;
  trend: number | null;
  period: string;
}

// ══════════════════════════════════════════════════════════
//  Audit Trail
// ══════════════════════════════════════════════════════════

export interface AuditLogEntry {
  id: string;
  user_id: string | null;
  user_name: string | null;
  action: string;
  entity_type: string;
  entity_id: string | null;
  old_values: unknown | null;
  new_values: unknown | null;
  ip_address: string | null;
  module: string | null;
  description: string | null;
  created_at: string;
}

export interface AuditLogSummary {
  id: string;
  user_name: string | null;
  action: string;
  entity_type: string;
  entity_id: string | null;
  module: string | null;
  description: string | null;
  created_at: string;
}

export interface AccessLogEntry {
  id: string;
  user_id: string;
  user_name: string | null;
  entity_type: string;
  entity_id: string | null;
  patient_id: string | null;
  patient_name: string | null;
  action: string;
  ip_address: string | null;
  module: string | null;
  created_at: string;
}

export interface AuditLogQuery {
  module?: string;
  entity_type?: string;
  entity_id?: string;
  user_id?: string;
  action?: string;
  from?: string;
  to?: string;
  page?: number;
  per_page?: number;
}

export interface AccessLogQuery {
  patient_id?: string;
  user_id?: string;
  entity_type?: string;
  module?: string;
  from?: string;
  to?: string;
  page?: number;
  per_page?: number;
}

export interface AuditStats {
  total_entries: number;
  today_entries: number;
  top_modules: ModuleCount[];
  top_users: UserActionCount[];
  action_breakdown: ActionCount[];
}

export interface ModuleCount {
  module: string | null;
  count: number;
}

export interface UserActionCount {
  user_name: string | null;
  count: number;
}

export interface ActionCount {
  action: string;
  count: number;
}

export interface LogAccessRequest {
  entity_type: string;
  entity_id?: string;
  patient_id?: string;
  module?: string;
}

export interface IntegrityResult {
  valid: boolean;
  total_checked: number;
  broken_at: string | null;
  expected_hash: string | null;
  actual_prev_hash: string | null;
}
